### PR TITLE
Rebuild image on changes to static class variables

### PR DESCRIPTION
**Pull request by TheQuantumFractal**
## Describe your changes

Addresses MOD-2989

This code change allows for static class variables that are accessed by the build function to be tracked for rebuilds. In particular, it determines the static class variables of the parent class if one is defined. It then traverses the code of the function to see which class attributes are accessed. The set of accessed static class variables is then tracked alongside the globals in the build_function definition.

Also added tests (tests need Mock Client Servicer to only build an image on new requests for ImageGetOrCreate)

<details> <summary>Backward/forward compatibility checks</summary>

---

Check these boxes or delete any item (or this section) if not relevant for this PR.

- [-] Client+Server: this change is compatible with old servers
**NOTE: This change will force image rebuilds for all classes with static class variables**
- [x] Client forward compatibility: this change ensures client can accept data intended for later versions of itself

Note on protobuf: protobuf message changes in one place may have impact to
multiple entities (client, server, worker, database). See points above.

---

</details>

## Changelog

- Image layers defined with a `@modal.build` method will now rebuild when there are changes to the values of any *class variables* referenced by the build method.

### CODE REVIEW

**Comment by mwaskom** [comment]
Not a comment on the implementation, but IMO, this might produce a net decrease in the coherence of the client behavior. I think it's especially a little weird to rebuild the image on changes to static class variables — but not on changes to class initialization parameters — especially since init params are one of the main reasons to use the class Function pattern.

I can't see the Slack discussion that's linked from the Linear ticket though; @aksh-at do you remember the original context here?

File: test/image_test.py [hunk 1]
```diff
@@ -750,6 +750,30 @@ def f(self):
...
+
+    @build()
+    def build_func(self):
+        global VARIABLE_5
+
+        print("foo!", VARIABLE_5)
+        print("static var", FooInstance.bar)
+
+
+def test_image_cls_var_rebuild(client, servicer):
```
Comment (thundergolfer) [inline test/image_test.py:764]: Would be good to have a couple of the 'negative' cases covered.

1. not rebuilding when not changed
2. not rebuilding when static var not referenced

File: modal/_utils/function_utils.py [hunk 2]
```diff
@@ -190,6 +190,25 @@ def serialized_function(self) -> bytes:
...
+            cls_vars = {
+                attr: getattr(self.cls, attr)
+                for attr in dir(self.cls)
+                if not callable(getattr(self.cls, attr)) and not attr.startswith("__")
+            }
+            return cls_vars
+        return {}
+
+    def get_attrs(self) -> Dict[str, Any]:
+        from .._vendor.cloudpickle import _extract_code_attr
```
Comment (mwaskom) [inline modal/_utils/function_utils.py:204]: Does this import need to be delayed? There shouldn't be a circularity.

File: modal/_utils/function_utils.py [hunk 3]
```diff
@@ -190,6 +190,25 @@ def serialized_function(self) -> bytes:
...
+        if self.cls is not None:
+            cls_vars = {
+                attr: getattr(self.cls, attr)
+                for attr in dir(self.cls)
+                if not callable(getattr(self.cls, attr)) and not attr.startswith("__")
+            }
+            return cls_vars
+        return {}
+
+    def get_attrs(self) -> Dict[str, Any]:
```
Comment (mwaskom) [inline modal/_utils/function_utils.py:203]: Nit: should this be `get_cls_var_attrs` or something like that?

File: modal/_vendor/cloudpickle.py [hunk 4]
```diff
@@ -323,6 +323,9 @@ def _extract_code_globals(co):
 
     return out_names
 
+def _extract_code_attr(co):
+    out_names = {name: None for name in _walk_attr_ops(co)}
+    return out_names
```
Comment (mwaskom) [inline modal/_vendor/cloudpickle.py:328]: Hm, I don't think that we should be making changes in vendored files; it gets too confusing if we ever need to update them from the upstream source.

File: modal/_vendor/cloudpickle.py [hunk 5]
```diff
@@ -323,6 +323,9 @@ def _extract_code_globals(co):
 
     return out_names
 
+def _extract_code_attr(co):
```
Comment (thundergolfer) [inline modal/_vendor/cloudpickle.py:326]: These modifications essentially fork the vendored code. Can we move these out?

File: test/image_test.py [hunk 6]
```diff
@@ -750,6 +750,48 @@ def f(self):
...
+    rebuild_app = App()
+    image_id = -1
+    rebuild_app.cls(image=Image.debian_slim())(FooInstance)
+    with rebuild_app.run(client=client):
+        image_id = list(servicer.images.keys())[-1]
```
Comment (mwaskom) [inline test/image_test.py:770]: Not worth rewriting the code, but just an FYI, you can just do `list(my_dict)` to get a list of the keys, you don't need to call `.keys()` directly.

File: test/image_test.py [hunk 7]
```diff
@@ -750,6 +750,48 @@ def f(self):
         print("bar!", VARIABLE_6)
 
 
+class FooInstance:
+    bar: str = "5"
+    barbar: str = "2"
```
Comment (mwaskom) [inline test/image_test.py:755]: Slight suggestion to make the tests below easier to follow, could name the attributes something like `used_by_build_method`, `not_used_by_build_method`

**Review by mwaskom — CHANGES_REQUESTED** [review]
Mostly nits and suggestions, but I do think we should figure out a different home for your additional extraction utilities to reduce the chance for future confusion if we need to update the vendored cloudpickle.

File: test/image_test.py (hunk 6 above)
Comment (TheQuantumFractal) [reply test/image_test.py]: Done

File: test/image_test.py (hunk 7 above)
Comment (TheQuantumFractal) [reply test/image_test.py]: Done

File: modal/_vendor/cloudpickle.py (hunk 4 above)
Comment (TheQuantumFractal) [reply modal/_vendor/cloudpickle.py]: Moved everything into function_utils but things may be a little sloppier now (?)

File: modal/_utils/function_utils.py (hunk 3 above)
Comment (TheQuantumFractal) [reply modal/_utils/function_utils.py]: Done

File: modal/_utils/function_utils.py [hunk 8]
```diff
@@ -190,6 +190,33 @@ def serialized_function(self) -> bytes:
...
+                for attr in dir(self.cls)
+                if not callable(getattr(self.cls, attr)) and not attr.startswith("__")
+            }
+            return cls_vars
+        return {}
+
+    def get_cls_var_attrs(self) -> Dict[str, Any]:
+        import dis
+
+        import opcode
```
Comment (mwaskom) [inline modal/_utils/function_utils.py:206]: Are these heavy imports? We should already have loaded them anyway on client startup via `modal._serialization -> modal._vendor.cloudpickle`. Maybe we should care more about optimizing client startup time (although I think we need serialization utilities to do most things).

File: test/image_test.py [hunk 9]
```diff
@@ -750,9 +750,52 @@ def f(self):
...
+    rebuild_app = App()
+    image_id = -1
+    rebuild_app.cls(image=Image.debian_slim())(FooInstance)
+    with rebuild_app.run(client=client):
+        image_id = list(servicer.images)[-1]
+    FooInstance.used_by_build_method = "rebuild"
+    rebuild_app.cls(image=Image.debian_slim())(FooInstance)
+    with rebuild_app.run(client=client):
+        image_id2 = list(servicer.images)[-1]
+    FooInstance.used_by_build_method = "normal"
```
Comment (mwaskom) [inline test/image_test.py:775]: Did you intend to check that this doesn't rebuild _again_ after you revert the change? That seems like a good idea if so.

**Review by thundergolfer — APPROVED** [review]

File: modal/_utils/function_utils.py [hunk 10]
```diff
@@ -190,6 +190,33 @@ def serialized_function(self) -> bytes:
...
+
+        LOAD_ATTR = opcode.opmap["LOAD_ATTR"]
+        STORE_ATTR = opcode.opmap["STORE_ATTR"]
```
Comment (mwaskom) [inline modal/_utils/function_utils.py:209]: Is it weird to rebuild if users only _store_ a new value for a class attribute, but don't read it, during the build phase? (Actually, separate discussion, but since we're doing this anyay we should maybe warn or error if users try to set class/instance attributes in `@build`, since that won't propagate to runtime containers, and they get confused about it occasionally.)

File: modal/_utils/function_utils.py [hunk 11]
```diff
@@ -190,6 +190,33 @@ def serialized_function(self) -> bytes:
...
+        ATTR_OPS = (LOAD_ATTR, STORE_ATTR)
+
+        func = self.raw_f
+        code = func.__code__
+        f_attr_ops = {instr.argval for instr in dis.get_instructions(code) if instr.opcode in ATTR_OPS}
```
Comment (mwaskom) [inline modal/_utils/function_utils.py:214]: Just double-checking for my own understanding that `instr.argval` here will be the attribute *name* not attribute *value*.

**Review by mwaskom — APPROVED** [review]
Nice!

File: modal/_utils/function_utils.py (hunk 11 above)
Comment (TheQuantumFractal) [reply modal/_utils/function_utils.py:214]: Yep. It's going to be the name of the attribute accessed in the opcode

File: test/image_test.py (hunk 9 above)
Comment (TheQuantumFractal) [reply test/image_test.py:775]: Oh sure good point

File: modal/_utils/function_utils.py (hunk 8 above)
Comment (TheQuantumFractal) [reply modal/_utils/function_utils.py:206]: They should be pretty light

File: modal/_utils/function_utils.py (hunk 10 above)
Comment (TheQuantumFractal) [reply modal/_utils/function_utils.py:209]: Oh this is fair

File: modal/_utils/function_utils.py [hunk 12]
```diff
@@ -190,6 +190,42 @@ def serialized_function(self) -> bytes:
...
+        func = self.raw_f
+        code = func.__code__
+        f_attr_ops = set()
+        for instr in dis.get_instructions(code):
+            if instr.opcode == LOAD_ATTR:
+                f_attr_ops.add(instr.argval)
+            elif instr.opcode == STORE_ATTR:
+                f_attr_ops.add(instr.argval)
+                logger.warning(
+                    "\n\nWarning: %s set in @build. You must set class variables in "
```
Comment (TheQuantumFractal) [inline modal/_utils/function_utils.py:220]: @mwaskom does this warning seem ok?

File: modal/_utils/function_utils.py (hunk 8 above)
Comment (mwaskom) [reply modal/_utils/function_utils.py:206]: Cool, not reason not to import them at the top of the module then IMO

File: modal/_utils/function_utils.py (hunk 12 above)
Comment (mwaskom) [reply modal/_utils/function_utils.py:220]: I think we'd probably want to issue it via the `warning` module — but since it's not directly related let's defer this change and think about it a tiny bit to make sure it's a good idea. It makes sense, but could potentially be annoying if we don't do it right.

Comment (TheQuantumFractal) [reply modal/_utils/function_utils.py]: Ok I'll remove it for now. Are we otherwise good to merge? This change will force rebuilds for all users that have static variables in their class so I do want to make sure things are ok on that front

Comment (mwaskom) [reply modal/_utils/function_utils.py]: > This change will force rebuilds for all users that have static variables in their class so I do want to make sure things are ok on that front.

Ah that's a great callout — we should be a little conservative about that.

On the one hand, layer defined by the class `@build` function should be the final layer(s) so won't trigger cascading  reinstalls of packages, which is often where people get into trouble (if they don't pin their dependency versions). On the other hand, they're often used for doing things like downloading models, which will take a while (and in some rare cases, break, if people point at a model revision that gets removed).

I guess it should specifically be limited to apps with build functions *that read or write to class variables*, is that right? In that case, I suspect it's a narrow-enough population who will directly benefit going forward.

Comment (TheQuantumFractal) [reply modal/_utils/function_utils.py]: This is correct. It's only apps with build functions that read / write to class variables. I don't think that's a large population, but good to flag in any case.
